presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewCandidateExtractor.java

```diff
@@ -33,38 +37,37 @@
     private final Metadata metadata;
     private final Session session;
 
+    Set<QualifiedObjectName> withTables = new HashSet<>();
+
     public MaterializedViewCandidateExtractor(Session session, Metadata metadata)
     {
         this.session = requireNonNull(session, "session is null");
         this.metadata = requireNonNull(metadata, "metadata is null");
     }
 
+    @Override
+    protected Void visitWithQuery(WithQuery node, Void context)
+    {
+        withTables.add(createQualifiedObjectName(session, node, QualifiedName.of(node.getName().toString())));
+        return super.visitWithQuery(node, null);
+    }
+
     @Override
     protected Void visitTable(Table node, Void context)
     {
         tableNames.add(createQualifiedObjectName(session, node, node.getName()));
         return null;
     }
 
-    public Set<QualifiedObjectName> getMaterializedViewCandidates()
+    public Map<QualifiedObjectName, List<QualifiedObjectName>> getMaterializedViewCandidatesForTable()
     {
-        Set<QualifiedObjectName> materializedViewCandidates = new HashSet<>();
+        Map<QualifiedObjectName, List<QualifiedObjectName>> baseTableToMaterializedViews = new HashMap<>();
```

Comment:
I would suggest using `ImmutableMap` instead of `HashMap`. We favor immutable data structures where possible.